[FLINK-2678]DataSet API does not support multi-dimensional arrays as keys

sbcd90 opened the pull request:
Hello,

@tillrohrmann I have added support for multi-dimensional arrays as keys in Dataset api. Please review & merge.

sbcd90 commented:
@StephanEwen could you also help reviewing the code?

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java:
```diff
@@ -0,0 +1,205 @@
+import java.lang.reflect.Array;
+import java.util.Arrays;
+
+public class GenericArrayComparator<T> extends TypeComparator<T[]> implements java.io.Serializable {
```
Java docs missing

tillrohrmann replied:
Why not calling it `ObjectArrayComparator`?

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java:
```diff
@@ -0,0 +1,205 @@
+		return Arrays.hashCode(record);
+	}
+
+	private int compareValues(Object first, Object second) {
```
This only works for primitive types. What about complex types?

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java:
```diff
@@ -0,0 +1,205 @@
+	}
+
+	@Override
+	public int compare(T[] first, T[] second) {
```
I think the compare method should only check that the length of `first` and `second` are the same and then apply the component comparator (derived from `ObjectArrayTypeInfo.componentInfo`) on the elements.

sbcd90 replied:
`ObjectArrayTypeInfo.componentInfo` is of type `TypeInformation<T>` & accessing a method like `createComparator` from one of the implementing classes of `TypeInformation<T>` is difficult. One possible solution is to have a static map like `TYPES` which maps classes to comparator classes. From `ObjectArrayTypeInfo.componentInfo` I get the typeClass & access the corresponding comparator from the map. Is there any other better solution?

tillrohrmann replied:
You should check whether `componentInfo` is an `AtomicType` or a `ComponentType` and then create the comparators via the provided methods. You can take a look at the `ComponentType.createComparator` implementation.

tillrohrmann commented on flink-core/src/test/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparatorTest.java:
```diff
@@ -0,0 +1,82 @@
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.junit.Assert;
+
+public class GenericArrayComparatorTest extends ComparatorTestBase<char[][]> {
```
Non primitive types should be tested as well.

sbcd90 replied:
Should another test class be created for testing non-primitive types?

tillrohrmann commented on flink-java/src/test/java/org/apache/flink/api/java/operator/DataSinkTest.java:
```diff
@@ -311,7 +311,7 @@ public void testPojoSingleOrderFull() {
 
-	@Test(expected = InvalidProgramException.class)
+	@Test
 	public void testArrayOrderFull() {
```
That is not how you fix a test. When the behaviour of the program changes, then one should also adapt the test accordingly. E.g. checking that the order of the elements is correct.

tillrohrmann commented:
Thanks for your contribution @sbcd90. I think the PR needs still a bit of work before it can be merged. The implementation of `GenericArrayComparator` only works for primitive types. Furthermore, I think that it is easier to construct the comparator out of the component type information of the `ObjectArrayTypeInfo` instead of doing the type analysis again every time you call `compare`.

sbcd90 commented:
@tillrohrmann Thanks for the review. I had a few questions based on your comments. Kindly help me in getting the questions answered so that I can proceed.

sbcd90 commented:
Hello @tillrohrmann  ..I have made all the changes you have mentioned.

- support for CompositeType.
- added test cases for both primitive type & composite type
- modified the changed test case in DataSinkTest.java
- changed the name of Comparator to ObjectArrayComparator.java
- javadocs are auto-generated by intellij.

Please please please please review my code now. Kindly let me know if there are any further changes required. Kindly let me know if this can be merged now also.

sbcd90 commented:
Hello @tillrohrmann please review the commit as the work to be done is nearly complete I believe. Please comment.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/api/java/typeutils/ObjectArrayTypeInfo.java:
```diff
@@ -72,15 +77,59 @@ public int getTotalFields() {
+	}
+
+	@SuppressWarnings("unchecked")
+	private TypeComparator<? super Object> getBaseComparatorInfo(TypeInformation<? extends Object> componentInfo, boolean sortOrderAscending, ExecutionConfig executionConfig) {
```
Why are you extracting for certain types the element comparator (for example the ObjectArrayTypeInfo) and for other you create the `TypeComparator` (for example the `CompositeTypeInfo`)? I don't get it. Why do you need the `getBaseComparatorInfo` method at all? Simply check in `createComparator` the different subtypes and then create the `TypeComparator`.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/api/java/typeutils/ObjectArrayTypeInfo.java:
```diff
@@ -72,15 +77,59 @@ public int getTotalFields() {
+
+		return (TypeComparator<T>) new ObjectArrayComparator<T,C>(
+			sortOrderAscending,
+			(GenericArraySerializer<T>) createSerializer(executionConfig),
```
Why this cast here?

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/ObjectArrayComparator.java:
```diff
@@ -0,0 +1,199 @@
+	}
+
+	@SuppressWarnings("unchecked")
+	private int parseGenericArray(Object firstArray, Object secondArray) {
```
Why exactly are you doing all the array operation on `Objects`?

tillrohrmann replied:
I think this is not the right way to go. Simply check then length of the arrays in the `compare` method and then call for all the elements the type comparator of the element type of the array.

tillrohrmann commented on flink-core/src/test/java/org/apache/flink/api/java/typeutils/runtime/ObjectArrayComparatorCompositeTypeTest.java:
```diff
@@ -0,0 +1,106 @@
+		((Tuple2<String, Integer>[][][]) result)[0][0][0].f0 = "be";
+		((Tuple2<String, Integer>[][][]) result)[0][0][0].f1 = 2;
+
+		((Tuple2<String, Integer>[][][]) result)[0][1][0] = new Tuple2<String, Integer>();
```
you can write something like `Tuple2.of("not", 3)`. Makes your life easier.

tillrohrmann commented:
I think the PR is still not in a good shape. I pointed out the problematic code sections.

Also please bear in mind that people will always try to review your code as soon as possible. But sometimes people are just busy and it takes some time for them to review a PR. In such a case, one simply has to be a little bit patient with them.
